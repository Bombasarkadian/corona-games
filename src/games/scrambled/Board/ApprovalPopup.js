import React from "react";
import PropTypes from "prop-types";
import { useTranslation } from "react-i18next";
import { Icon, Button } from "semantic-ui-react";
import InfoPopup from "./InfoPopup";

const propTypes = {
  popupHandleRef: PropTypes.object,
  open: PropTypes.bool,
  errors: PropTypes.arrayOf(PropTypes.string),
  newWords: PropTypes.arrayOf(PropTypes.object),
  specialBonus: PropTypes.bool,
  onReject: PropTypes.func,
  onApprove: PropTypes.func,
};

const ApprovalPopup = ({
  popupHandleRef,
  open,
  errors,
  newWords,
  specialBonus,
  onReject,
  onApprove,
}) => {
  const { t } = useTranslation("scrambled");

  return (
    <InfoPopup clickable {...{ popupHandleRef, open, errors, newWords, specialBonus }}>
      <Button color="red" onClick={onReject}>
        <Icon name="close" /> {t("game.modals.approval.reject")}
      </Button>
      <Button color="green" onClick={onApprove}>
        <Icon name="checkmark" /> {t("game.modals.approval.approve")}
      </Button>
    </InfoPopup>
  );
};

ApprovalPopup.propTypes = propTypes;
export default ApprovalPopup;
