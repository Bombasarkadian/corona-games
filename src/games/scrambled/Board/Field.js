import React from "react";
import PropTypes from "prop-types";
import { Icon } from "semantic-ui-react";
import { useTranslation } from "react-i18next";

import Tile from "../Tile";
import { BONUSES } from "../config";

import styles from "./Field.module.scss";
import clsx from "clsx";

const dimmerPropTypes = {
  enabled: PropTypes.bool,
};
const Dimmer = ({ enabled }) => (
  <div
    className={clsx({
      [styles.field]: true,
      [styles.dimmer]: true,
      [styles.dimmer_enabled]: enabled,
    })}
  ></div>
);
Dimmer.propTypes = dimmerPropTypes;

const fieldPropTypes = {
  base: PropTypes.object.isRequired,
  overlay: PropTypes.object,
  clickable: PropTypes.bool,
  handleFieldClick: PropTypes.func.isRequired,
  selectionEnabled: PropTypes.bool,
};
const Field = ({ base, overlay, clickable, handleFieldClick, selectionEnabled }) => {
  const { t } = useTranslation("scrambled");

  // Permanently placed letter
  if (base.letter || base.replacement) {
    return (
      <div className={styles.field}>
        <Tile {...base} />
      </div>
    );
  }

  // Empty field with a bonus
  if (base.bonus) {
    const content = base.start ? (
      <Icon name="star" size="big" disabled fitted />
    ) : (
      <div className={styles.bonus}>
        <div className={styles.bonus_multiply}>{base.bonus.multiply}x</div>
        <div className={styles.bonus_type}>{t(`game.info.bonuses.${base.bonus.type}`)}</div>
      </div>
    );
    return (
      <div
        className={clsx({ [styles.field]: true, [styles.field_clickable]: clickable })}
        style={{
          background: BONUSES[base.bonus.type][base.bonus.multiply],
        }}
        onClick={handleFieldClick}
        ref={overlay ? overlay.popupRef : undefined}
      >
        {overlay ? <Tile highlighted bonus={base.bonus} {...overlay} /> : content}
        <Dimmer enabled={selectionEnabled && !clickable} />
      </div>
    );
  }

  // Empty field
  return (
    <div
      className={clsx({ [styles.field]: true, [styles.field_clickable]: clickable })}
      onClick={handleFieldClick}
      ref={overlay ? overlay.popupRef : undefined}
    >
      {overlay ? <Tile highlighted {...overlay} /> : <></>}
      <Dimmer enabled={selectionEnabled && !clickable} />
    </div>
  );
};
Field.propTypes = fieldPropTypes;

export default Field;
