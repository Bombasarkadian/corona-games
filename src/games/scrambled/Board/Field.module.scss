.field {
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  text-align: center;
  font-size: 12px;
  line-height: 1;
  color: #ffffff;
  border-radius: 5px;
  border: 0px solid #ffffff;
  background: #ffffff;
  font-weight: 500;
  cursor: auto;

  &_clickable {
    .dimmer {
      cursor: pointer;
    }
  }

  .dimmer {
    position: absolute;
    width: 100%;
    height: 100%;
    background: #000000;
    opacity: 0;
    transition: opacity 0.3s;

    &_enabled {
      opacity: 0.15;
    }
  }

  & .bonus {
    font-weight: 600;

    &_multiply {
      margin-bottom: 3px;
      font-size: 12px;
    }

    &_type {
      text-transform: uppercase;
      font-size: 8px;
    }
  }
}
