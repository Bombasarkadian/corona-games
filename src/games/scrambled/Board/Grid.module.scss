.grid {
  background: #717171;
  user-select: none;
  display: flex;
  align-items: center;
  justify-content: center;
  width: calc(18.5 * 46px);
  height: calc(18.5 * 46px);
  border-radius: 3px;

  & table {
    border: 2px solid #e4e4e4;
    border-radius: 5px;
    background: #e4e4e4;
    box-shadow: rgba(0, 0, 0, 0.4) 0px 0px 10px 2px;
    border-spacing: 5px;
    transition: background-color 0.3s 0.08s, border-color 0.3s 0.08s;
  }

  .cell {
    width: 46px;
    height: 46px;
    padding: 0;
    position: relative;
  }

  &_selection {
    & table {
      border-color: #b1b1b1;
      background: #b1b1b1;
    }
  }
}
