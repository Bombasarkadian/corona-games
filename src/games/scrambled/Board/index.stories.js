import React from "react";

import { scrambledDecorator } from "../GameContextMock";
import Board from ".";

export default {
  component: Board,
  title: "Scrambled/Board",
  excludeStories: /.*Data$/,
  decorators: [scrambledDecorator],
};

export const Default = () => <Board />;
