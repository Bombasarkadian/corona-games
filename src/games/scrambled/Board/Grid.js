import React from "react";
import PropTypes from "prop-types";

import Field from "./Field";

import styles from "./Grid.module.scss";
import clsx from "clsx";

const propTypes = {
  board: PropTypes.array.isRequired,
  playerTiles: PropTypes.array.isRequired,
  clickable: PropTypes.func.isRequired,
  handleFieldClick: PropTypes.func.isRequired,
  selectionEnabled: PropTypes.bool,
};
const Grid = ({ board, playerTiles, clickable, handleFieldClick, selectionEnabled }) => (
  <div className={clsx({ [styles.grid]: true, [styles.grid_selection]: selectionEnabled })}>
    <table>
      <tbody>
        {board.map((rowContainer, y) => {
          return (
            <tr key={y} style={{ height: 46 }}>
              {rowContainer.row.map((el, x) => {
                return (
                  <td key={`${x},${y}`} className={styles.cell}>
                    <Field
                      base={el === 0 ? {} : el}
                      overlay={
                        playerTiles && playerTiles.find((tile) => tile.x === x && tile.y === y)
                      }
                      handleFieldClick={() => handleFieldClick(x, y)}
                      clickable={clickable(x, y)}
                      selectionEnabled={selectionEnabled}
                    />
                  </td>
                );
              })}
            </tr>
          );
        })}
      </tbody>
    </table>
  </div>
);
Grid.propTypes = propTypes;

export default Grid;
