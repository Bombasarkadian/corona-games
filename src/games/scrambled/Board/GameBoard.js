import React, { useState, useEffect, useCallback, useRef } from "react";
import { useTranslation } from "react-i18next";
import { Header, Confirm, Sticky } from "semantic-ui-react";

import { useBoardGame } from "contexts/BoardGameContext";
import GameLayout from "components/layout/GameLayout";

import Grid from "./Grid";
import TilesPanel from "./TilesPanel";
import InfoPopup from "./InfoPopup";
import { remove, orderBy, pick } from "lodash";
import { availableLaguages } from "../data/tiles";
import { tilesPlacementErrors, newWords, filterPlayedTiles, canPlaceTile } from "../utils";
import ApprovalModal from "../Modals/ApprovalModal";
import SwapTilesModal from "../Modals/SwapTilesModal";
import BlankTileModal from "../Modals/BlankTileModal";

const GameBoard = () => {
  const {
    G,
    ctx: { phase, currentPlayer },
    player: { id, tiles, stage, isCurrentPlayer },
    moves: { PlayTiles, SwapTiles, SkipTurn, Approve },
    chatMessages,
    sendChatMessage,
  } = useBoardGame();

  const { t } = useTranslation("scrambled");
  const [selectedTile, setSelectedTile] = useState(null);
  const [playerTiles, setPlayerTiles] = useState([]);
  const [moveErrors, setMoveErrors] = useState([]);
  const [selectedBlankTile, setSelectedBlankTile] = useState(null);
  const [blankTileReplacement, setBlankTileReplacement] = useState(null);
  const [swapTilesModal, setSwapTilesModal] = useState(false);
  const [skipTurnModal, setSkipTurnModal] = useState(false);
  const [selectedForSwap, setSelectedForSwap] = useState([]);
  const [popupOpen, setPopupOpen] = useState(false);
  const [previewTiles, setPreviewTiles] = useState(null);
  const popupHandleRef = useRef();
  const stickyRef = useRef();

  const currentLanguage = availableLaguages.find(({ key }) => key === G.language);
  const playedTiles = filterPlayedTiles(playerTiles);
  const canMakeMove = isCurrentPlayer && phase === "play" && stage === "play";
  const canReturnTiles = playedTiles.length > 0;
  const canPlayTiles = playedTiles.length > 0 && moveErrors.length === 0;
  const canSwapTiles = G.tilesLeft >= 7;
  const approvalPhase = phase === "play" && stage === "approve";

  const onReturnTiles = useCallback(() => {
    Object.values(playerTiles).forEach((tile) => {
      delete tile.x;
      delete tile.y;
      delete tile.popupRef;
    });
    setPlayerTiles([...playerTiles]);
    setPopupOpen(false);
  }, [playerTiles]);

  const onPlayTiles = useCallback(() => {
    playerTiles.forEach((tile) => {
      tile.popupRef = undefined;
    });
    setPopupOpen(false);
    PlayTiles({ tiles: playerTiles });
  }, [PlayTiles, playerTiles]);

  const onSkipTurn = useCallback(() => {
    setSkipTurnModal(false);
    playerTiles.forEach((tile) => {
      tile.popupRef = undefined;
    });
    setPopupOpen(false);
    SkipTurn();
  }, [SkipTurn, playerTiles]);

  const onSelectForSwap = useCallback(
    (tile) => {
      selectedForSwap.includes(tile)
        ? remove(selectedForSwap, (el) => el === tile)
        : selectedForSwap.push(tile);
      setSelectedForSwap([...selectedForSwap]);
    },
    [selectedForSwap]
  );

  useEffect(() => {
    if (tiles !== undefined) setPlayerTiles(tiles);
  }, [tiles]);

  useEffect(() => {
    // player.isCurrentPlayer didn't seem to work properly
    if (id !== undefined && id.toString() !== currentPlayer) {
      setMoveErrors(["invalid_player"]);
      return;
    }
    const playedTiles = filterPlayedTiles(playerTiles);
    setMoveErrors(tilesPlacementErrors(G, currentPlayer, playedTiles));
  }, [G, currentPlayer, id, playerTiles]);

  useEffect(() => {
    if (canMakeMove || chatMessages.length <= 0) return;
    const lastMessage = chatMessages[chatMessages.length - 1].payload;

    if (lastMessage.type && lastMessage.type.startsWith("MovePreview")) {
      setPreviewTiles(lastMessage.data.map((el) => ({ preview: true, ...el })));
    }
  }, [canMakeMove, chatMessages]);

  const selectTile = useCallback(
    (tile) => {
      if (!canMakeMove) return;
      if (selectedTile === tile) setSelectedTile(null);
      else setSelectedTile(tile);
    },
    [canMakeMove, selectedTile]
  );

  const canBeClicked = useCallback(
    (x, y) => {
      // Player can't click outside of their turn
      if (!isCurrentPlayer) return false;
      // Player can always click on a temporarily placed tile to remove it
      if (playerTiles.find((tile) => tile.x === x && tile.y === y)) return true;
      if (selectedTile !== null) return canPlaceTile(G, playedTiles, x, y);
      return false;
    },
    [G, isCurrentPlayer, playedTiles, playerTiles, selectedTile]
  );

  const clickBoard = useCallback(
    (x, y) => {
      if (!canBeClicked(x, y)) return;
      setPreviewTiles(null);

      // Remove a tile that's on x,y
      const currentIdx = playerTiles.findIndex((tile) => tile.x === x && tile.y === y);
      if (currentIdx >= 0) {
        delete playerTiles[currentIdx].x;
        delete playerTiles[currentIdx].y;
      }

      // Place new tile on x,y
      if (selectedTile !== null) {
        if (!selectedTile.letter) {
          setSelectedBlankTile(selectedTile);
        }
        selectedTile.x = x;
        selectedTile.y = y;
      }

      setSelectedTile(null);
      // Force re-rendering of placed tiles
      setPlayerTiles([...playerTiles]);

      playerTiles.forEach((tile) => {
        tile.popupRef = undefined;
      });

      const playedTiles = filterPlayedTiles(playerTiles);
      // Select new tile for holding info popup
      if (playedTiles.length > 0)
        orderBy(playedTiles, ["y", "x"], ["asc", "asc"])[0].popupRef = popupHandleRef;
      setPopupOpen(playedTiles.length > 0);

      if (G.preview)
        sendChatMessage({
          type: "MovePreview",
          data: playedTiles.map((el) => pick(el, ["x", "y"])),
        });
    },
    [G.preview, canBeClicked, playerTiles, selectedTile, sendChatMessage]
  );

  const extraPlayerContent = useCallback(
    ({ tilesCount }) => (
      <>
        {t("game.info.tiles.owned")}: {tilesCount}
      </>
    ),
    [t]
  );

  return (
    <div ref={stickyRef}>
      <GameLayout
        gameName={t("game.name")}
        header={
          <div style={{ marginBottom: "14px" }}>
            <Header as="h2" textAlign="center">
              {t(`header.${phase}.${stage}`)}
              <Header.Subheader>{t(`subheader.${phase}.${stage}`)}</Header.Subheader>
            </Header>
            <Sticky context={stickyRef}>
              <TilesPanel
                disabled={!canMakeMove}
                onSwapTiles={() => setSwapTilesModal(true)}
                onSkipTurn={() => setSkipTurnModal(true)}
                tilesLeft={G.tilesLeft}
                {...{
                  playerTiles,
                  selectTile,
                  selectedTile,
                  onReturnTiles,
                  onPlayTiles,
                  canReturnTiles,
                  canSwapTiles,
                  canPlayTiles,
                }}
              />
            </Sticky>
          </div>
        }
        extraPlayerContent={extraPlayerContent}
      >
        <InfoPopup
          open={popupOpen && !selectedBlankTile && !swapTilesModal}
          errors={moveErrors}
          newWords={newWords(G.board, playerTiles)}
          specialBonus={playedTiles.length === 7}
          {...{ popupHandleRef }}
        />
        <Grid
          board={G.board}
          clickable={canBeClicked}
          handleFieldClick={clickBoard}
          playerTiles={stage === "wait_for_approval" ? G.pendingTiles : previewTiles || playerTiles}
          selectionEnabled={selectedTile !== null}
        />
        {G.pendingTiles && approvalPhase && (
          <ApprovalModal
            newWords={newWords(G.board, G.pendingTiles)}
            onReject={() => Approve(false)}
            onApprove={() => Approve(true)}
          />
        )}
        {swapTilesModal && (
          <SwapTilesModal
            onClose={() => {
              setSelectedForSwap([]);
              setSwapTilesModal(false);
            }}
            onSwap={() => {
              SwapTiles(selectedForSwap);
              setSwapTilesModal(false);
            }}
            {...{ playerTiles, onSelectForSwap, selectedForSwap }}
          />
        )}
        {selectedBlankTile !== null && (
          <BlankTileModal
            letters={currentLanguage.letters}
            onClose={() => {
              selectedBlankTile.replacement = blankTileReplacement;
              setPlayerTiles([...playerTiles]);
              setSelectedBlankTile(null);
            }}
            {...{ setBlankTileReplacement, blankTileReplacement }}
          />
        )}
        <Confirm
          open={skipTurnModal}
          onCancel={() => setSkipTurnModal(false)}
          onConfirm={onSkipTurn}
          cancelButton={t("game.modals.skip.cancel")}
          confirmButton={t("game.modals.skip.confirm")}
          content={t("game.modals.skip.content")}
        />
      </GameLayout>
    </div>
  );
};

export default GameBoard;
