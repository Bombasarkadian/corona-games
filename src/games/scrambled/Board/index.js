import React, { useState, useEffect, useCallback } from "react";
import { useTranslation } from "react-i18next";
import { Label, Button, Segment, Icon, Header } from "semantic-ui-react";

import { useBoardGame } from "contexts/BoardGameContext";
import GameLayout from "components/layout/GameLayout";

import Tile from "../Tile";
import Grid from "./Grid";
import SettingsBoard from "./SettingsBoard";

const Board = () => {
  const {
    G,
    ctx: { phase },
    player: { letters, stage },
    moves: { StartGame, PlayTiles },
  } = useBoardGame();

  const { t } = useTranslation("scrambled");
  const [selectedLetterIdx, setSelectedLetterIdx] = useState(null);
  const [playerLetters, setPlayerLetters] = useState([]);

  const hasGameStarted = phase === "play";

  useEffect(() => {
    if (letters !== undefined) setPlayerLetters(letters);
  }, [letters]);

  const selectLetter = useCallback(
    (idx) => {
      if (playerLetters[idx].x !== undefined && playerLetters[idx].y !== undefined) return;

      if (selectedLetterIdx === idx) setSelectedLetterIdx(null);
      else setSelectedLetterIdx(idx);
    },
    [playerLetters, selectedLetterIdx, setSelectedLetterIdx]
  );

  const clickBoard = useCallback(
    (x, y) => {
      // Remove a tile that's on x,y
      const current = playerLetters.find((tile) => tile.x === x && tile.y === y);
      if (current) {
        delete current.x;
        delete current.y;
      }

      // Place new tile on x,y
      if (selectedLetterIdx !== null) {
        playerLetters[selectedLetterIdx].x = x;
        playerLetters[selectedLetterIdx].y = y;
      }

      setSelectedLetterIdx(null);
      // Force re-rendering of placed tiles
      setPlayerLetters([...playerLetters]);
    },
    [selectedLetterIdx, setSelectedLetterIdx, playerLetters, setPlayerLetters]
  );

  const onReturnLetters = useCallback(() => {
    Object.values(playerLetters).forEach((tile) => {
      delete tile.x;
      delete tile.y;
    });
    setPlayerLetters([...playerLetters]);
  }, [playerLetters, setPlayerLetters]);

  const onPlayTiles = useCallback(() => {
    PlayTiles(playerLetters);
  }, [PlayTiles, playerLetters]);

  const canBeClicked = (x, y) => {
    if (selectedLetterIdx !== null) return !G.board[y].row[x].letter;
    if (playerLetters.find((tile) => tile.x === x && tile.y === y)) return true;
    return false;
  };

  if (!hasGameStarted) {
    return (
      <GameLayout
        gameName={t("game.name")}
        header={
          <Header as="h2" textAlign="center">
            {t(`header.${phase}`)}
            <Header.Subheader>{t(`subheader.${phase}.${stage}`)}</Header.Subheader>
          </Header>
        }
      >
        <SettingsBoard
          modes={G.modes}
          defaultMode={G.mode}
          styles={G.styles}
          defaultStyle={G.style}
          defaultPicturesCount={G.picturesCount}
          onStartGame={StartGame}
        />
      </GameLayout>
    );
  }

  return (
    <GameLayout
      gameName={t("game.name")}
      header={
        <>
          <Header as="h2" textAlign="center">
            {t(`header.${phase}.${stage}`)}
            <Header.Subheader>{t(`subheader.${phase}.${stage}`)}</Header.Subheader>
          </Header>
          <Segment disabled={phase !== "play" || stage !== "play"}>
            <Label as="span" color="green" ribbon>
              <Icon name="star" />
              Your tiles
            </Label>
            <div style={{ display: "flex", paddingTop: 10 }}>
              {playerLetters &&
                playerLetters.map((letter, idx) => (
                  <Tile
                    separate
                    key={idx}
                    onClick={() => selectLetter(idx)}
                    raised={selectedLetterIdx === idx}
                    disabled={letter.x !== undefined && letter.y !== undefined}
                    {...letter}
                  />
                ))}
              <Button color="red" onClick={onReturnLetters}>
                <Icon name="undo" />
                Return letters
              </Button>
              <Button color="green" onClick={onPlayTiles}>
                <Icon name="check" />
                Submit
              </Button>
            </div>
          </Segment>
        </>
      }
    >
      <Grid
        board={G.board}
        clickable={canBeClicked}
        handleFieldClick={clickBoard}
        playerLetters={playerLetters}
        selectionEnabled={selectedLetterIdx !== null}
      />
    </GameLayout>
  );
};

export default Board;
