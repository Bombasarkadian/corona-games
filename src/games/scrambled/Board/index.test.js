import React from "react";
import renderer from "react-test-renderer";
import Board from "./index";
import BoardGameContextMock from "../GameContextMock";
import { MemoryRouter } from "react-router";

describe("scrambled Board component", () => {
  it("renders correctly", () => {
    expect.hasAssertions();
    const component = renderer.create(
      <MemoryRouter>
        <BoardGameContextMock>
          <Board />
        </BoardGameContextMock>
      </MemoryRouter>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
