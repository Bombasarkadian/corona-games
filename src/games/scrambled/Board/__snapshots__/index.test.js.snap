// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrambled Board component renders correctly 1`] = `
<div>
  <div
    className="ui container"
  >
    <div
      className="mainMenu"
    >
      <div
        className="ui container"
      >
        <a
          className="logo"
          href="/"
          onClick={[Function]}
        >
          <img
            className="ui image"
            slt="Octoboard logo"
            src="/images/octoboard-small.svg"
          />
        </a>
        <span
          className="gameName"
        >
          game.name
        </span>
        <span
          className="ui small label"
          onClick={[Function]}
          style={Object {}}
        >
          <i
            aria-hidden="true"
            className="open lock icon"
            onClick={[Function]}
          />
          <div
            className="detail"
          >
            game.public
          </div>
        </span>
      </div>
    </div>
    <div
      className="ui container"
    >
      <div
        style={
          Object {
            "marginBottom": "14px",
          }
        }
      >
        <h2
          className="ui center aligned header"
        >
          header.play.wait
          <div
            className="sub header"
          >
            subheader.play.wait
          </div>
        </h2>
        <div
          className=""
        >
          <div
            style={Object {}}
          />
          <div
            className="ui sticky"
          >
            <div
              className="ui disabled segment"
            >
              <span
                className="ui green ribbon label"
                onClick={[Function]}
              >
                <i
                  aria-hidden="true"
                  className="star icon"
                  onClick={[Function]}
                />
                panel.tiles
              </span>
              <div
                style={
                  Object {
                    "display": "flex",
                    "paddingTop": 10,
                  }
                }
              >
                <button
                  className="ui red disabled button"
                  disabled={true}
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  <i
                    aria-hidden="true"
                    className="undo icon"
                    onClick={[Function]}
                  />
                  panel.actions.return
                </button>
                <button
                  className="ui orange disabled button"
                  disabled={true}
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  <i
                    aria-hidden="true"
                    className="refresh icon"
                    onClick={[Function]}
                  />
                  panel.actions.swap
                </button>
                <button
                  className="ui yellow disabled button"
                  disabled={true}
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  <i
                    aria-hidden="true"
                    className="forward icon"
                    onClick={[Function]}
                  />
                  panel.actions.skip
                </button>
                <button
                  className="ui green disabled button"
                  disabled={true}
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  <i
                    aria-hidden="true"
                    className="check icon"
                    onClick={[Function]}
                  />
                  panel.actions.play
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ui grid"
    >
      <div
        className="twelve wide computer sixteen wide mobile column"
      >
        <div
          className="row"
        >
          <div
            className="grid"
          >
            <table>
              <tbody>
                <tr
                  style={
                    Object {
                      "height": 46,
                    }
                  }
                >
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                      style={
                        Object {
                          "background": "#BB5C66",
                        }
                      }
                    >
                      <div
                        className="bonus"
                      >
                        <div
                          className="bonus_multiply"
                        >
                          3
                          x
                        </div>
                        <div
                          className="bonus_type"
                        >
                          word
                        </div>
                      </div>
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                    >
                      <div
                        className="tile"
                      >
                        <div
                          className="letter"
                        >
                          L
                        </div>
                        <div
                          className="points"
                        >
                          2
                        </div>
                      </div>
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                    >
                      <div
                        className="tile"
                      >
                        <div
                          className="letter"
                        >
                          O
                        </div>
                        <div
                          className="points"
                        >
                          1
                        </div>
                      </div>
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                    >
                      <div
                        className="tile"
                      >
                        <div
                          className="letter letter_replacement"
                        >
                          L
                        </div>
                        <div
                          className="points"
                        >
                          0
                        </div>
                      </div>
                    </div>
                  </td>
                </tr>
                <tr
                  style={
                    Object {
                      "height": 46,
                    }
                  }
                >
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                      style={
                        Object {
                          "background": "#21C4D8",
                        }
                      }
                    >
                      <div
                        className="bonus"
                      >
                        <div
                          className="bonus_multiply"
                        >
                          2
                          x
                        </div>
                        <div
                          className="bonus_type"
                        >
                          letter
                        </div>
                      </div>
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  style={
                    Object {
                      "height": 46,
                    }
                  }
                >
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                      style={
                        Object {
                          "background": "#16919F",
                        }
                      }
                    >
                      <div
                        className="bonus"
                      >
                        <div
                          className="bonus_multiply"
                        >
                          3
                          x
                        </div>
                        <div
                          className="bonus_type"
                        >
                          letter
                        </div>
                      </div>
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                      style={
                        Object {
                          "background": "#F9C54A",
                        }
                      }
                    >
                      <div
                        className="bonus"
                      >
                        <div
                          className="bonus_multiply"
                        >
                          2
                          x
                        </div>
                        <div
                          className="bonus_type"
                        >
                          word
                        </div>
                      </div>
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  style={
                    Object {
                      "height": 46,
                    }
                  }
                >
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                      style={
                        Object {
                          "background": "#F9C54A",
                        }
                      }
                    >
                      <div
                        className="bonus"
                      >
                        <div
                          className="bonus_multiply"
                        >
                          2
                          x
                        </div>
                        <div
                          className="bonus_type"
                        >
                          word
                        </div>
                      </div>
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                      style={
                        Object {
                          "background": "#F9C54A",
                        }
                      }
                    >
                      <i
                        aria-hidden="true"
                        className="star big disabled fitted icon"
                        onClick={[Function]}
                      />
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
      <div
        className="four wide computer sixteen wide mobile column"
      >
        <div
          className="row"
        >
          <h2
            className="ui center aligned header"
            style={
              Object {
                "marginBottom": "-5px",
              }
            }
          >
            game.players
            <span
              className="ui small label"
              onClick={[Function]}
              style={Object {}}
            >
              <i
                aria-hidden="true"
                className="open lock icon"
                onClick={[Function]}
              />
              <div
                className="detail"
              >
                game.public
              </div>
            </span>
          </h2>
          <div
            className="ui segments"
          />
          <div
            className="ui basic center aligned segment leaveSegment"
          >
            <button
              className="ui blue icon disabled left labeled button"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              tabIndex={-1}
            >
              <i
                aria-hidden="true"
                className="info circle icon"
                onClick={[Function]}
              />
              game.rules.header
            </button>
          </div>
          <div
            className="ui basic center aligned segment leaveSegment"
          >
            <button
              className="ui red icon left labeled button"
              onClick={[Function]}
            >
              <i
                aria-hidden="true"
                className="close icon"
                onClick={[Function]}
              />
              game.leave.button
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="wrapper"
    >
      <span
        className="triggerButton"
        data-unread={0}
        onClick={[Function]}
      />
      <div
        className="popupChat"
      >
        <div
          className="ui segments chat"
        >
          <div
            className="ui compact segment header"
          >
            game.chat.header
            <span
              className="closeIcon"
              onClick={[Function]}
            >
              <i
                aria-hidden="true"
                className="close icon"
                onClick={[Function]}
              />
            </span>
          </div>
          <div
            className="ui segment messages"
          >
            <div
              className="ui comments"
            />
          </div>
          <div
            className="ui segment inputBox"
          >
            <form
              onSubmit={[Function]}
            >
              <div
                className="ui fluid icon input"
              >
                <input
                  onChange={[Function]}
                  placeholder="game.chat.sendMessage"
                  type="text"
                  value=""
                />
                <i
                  aria-hidden="true"
                  className="send icon"
                  onClick={[Function]}
                />
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
    <div
      className="Toastify"
    />
  </div>
</div>
`;
